### PR TITLE
Split into even more processes

### PR DESCRIPTION
**Pull request by coolo**
Make the tests a process too - and make isotovideo an event loop.

This is heavily broken and unfinished - it's merely to give you an early glimpse :)

### CODE REVIEW

**Comment by okurz** [comment]
+1

**Comment by coolo** [comment]
I officially add code now ;) +559 −553

**Comment by coolo** [comment]
this works now - but this is such a heavy change that more people need to test it

File: isotovideo [hunk 1]
```diff
@@ -148,41 +147,28 @@ require $bmwqemu::vars{PRODUCTDIR} . "/main.pm";
 # set a default distribution if the tests don't have one
 $testapi::distri ||= distribution->new;
 
-# start the command fork before we get into the backend, the command child
-# is not supposed to talk to the backend directly
-$cpid = commands::start_server($bmwqemu::vars{QEMUPORT} + 1);
+testapi::init();
+
+use Data::Dumper;
+print Dumper($testapi::distri);
```
Comment (okurz) [inline isotovideo]: remove or use proper logging call?

File: isotovideo [hunk 2]
```diff
@@ -148,41 +147,28 @@ require $bmwqemu::vars{PRODUCTDIR} . "/main.pm";
...
 
-    if (!bmwqemu::alive) {
-        bmwqemu::start_vm or die $@;
-    }
+my ($testpid, $testfd) = autotest::start_process();
+
+init_backend();
+
+open(my $fd, ">", "os-autoinst.pid");
+print $fd "$$\n";
```
Comment (okurz) [inline isotovideo]: here, too

File: myjsonrpc.pm [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    my ($to_fd, $cmd) = @_;
+
+    # TODO: make this a class object
+    # allow regular expressions to be automatically converted into
+    # strings, using the Regex::TO_JSON function as defined at the end
+    # of this file.
+    my $JSON = JSON->new()->convert_blessed();
+    my $json = $JSON->encode($cmd);
+
+    #cluck "$$: send_json $json";
```
Comment (okurz) [inline myjsonrpc.pm]: delete

File: myjsonrpc.pm [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        $hash = $JSON->incr_parse();
+        if ($hash) {
+            # remember the trailing text
+            $sockets->{$fd} = $JSON->incr_text();
+            if ($hash->{QUIT}) {
+                print "received magic close\n";
+                return;
+            }
+            use Data::Dumper;
+            #bmwqemu::diag("read_json " . Dumper($hash));
```
Comment (okurz) [inline myjsonrpc.pm]: delete two lines

File: myjsonrpc.pm [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        $JSON->incr_parse($sockets->{$fd});
+        delete $sockets->{$fd};
+    }
+
+    my $s = IO::Select->new();
+    $s->add($socket);
+
+    my $hash;
+
+    # starting a IPMI host can take a while, so we need to be patient
```
Comment (okurz) [inline myjsonrpc.pm]: I think this line doesn't apply here anymore

**Comment by okurz** [comment]
Testing locally I did not manage to enable interactive mode, see

openqa_rework_interactive_does_not_stop

what I did:
start test run (gnome default scenario on SLES), clicked "interactive mode" as soon as grub showed up, wheel keeps on spinning.

See http://lord.arch/tests/1822 for logs.

**Comment by coolo** [comment]
I will have a look as time permits

**Comment by okurz** [comment]
On Monday 04 July 2016 22:08:34 Stephan Kulow wrote:
> I will have a look as time permits

As long as you don't merge "prematurely", there is time ;-)

**Comment by coolo** [comment]
the 1822 test is not running with the rework_interactive_mode openQA worker. Otherwise you see the annoying /isotovideo/status in the log (and the screenshot)

**Comment by coolo** [comment]
the /status lines disppear when I set the commands.pm webserver to warning instead of debug. But it's really handy at this point to know what commands were issued. And the annoying /status lines only appear in the log when you look at the live view

**Comment by okurz** [comment]
* I would also like to see the commands that have been issued as they appear because of manual interaction. But the continous automatic update is not helping. Maybe use loglevel "info" by default and output only the important events, i.e. not the one that happen all the item?

* Did you also test isotovideo call without being triggered by openQA?
* Does continuation from snapshots still work?
* how can I exit the needle editor w/o saving during live?
* clicking on "stop wait for needle" during run jumps to "#details" (in firefox?)
* http://gertrude/tests/213063#live clicked on "stop" but but stopped only after like 20 more needle matches. I pressed "stop" during grub menue but it only stopped after it typed user name and such in installer
* in interactive mode waiting for needle, when clicking on "disable interactive mode" exits interactive mode but VM is still stopped vs. expected: Should continue

**Comment by coolo** [comment]
14b0311 for your last point

**Comment by coolo** [comment]
https://github.com/os-autoinst/openQA/pull/767/commits/3a5f6a703ad1ccf96a78364215fbdb3f44da07db for the jumping buttons

**Comment by coolo** [comment]
b87da75 for annoying /status

All other points you mention are not regressions and exceed the scope of this PR

**Comment by okurz** [comment]
After we have been talking yesterday I want to answer one question by myself as reference for others:

> * how can I exit the needle editor w/o saving during live?

current approach is to either rely on browser "back" button or manually edit the adress in the adress bar to go back to the test details page. coolo wants to add a "Return without save" button to do this from the UI.

> * Did you also test isotovideo call without being triggered by openQA?

It seems I found at least one problem there. If isotovideo crashes at some point it might leave processes left behind, e.g. the command server.

> * Does continuation from snapshots still work?

@coolo ?

**Comment by coolo** [comment]
why would it not?

**Comment by okurz** [comment]
because we can't keep all eventualities in our head

**Comment by aaannz** [comment]
So far LGTM, the only thing is that this is once again update breaking openQA compatibility so needs to be updated in tandem. Can you add something like `isotovideo/version` route?

**Comment by coolo** [comment]
that means isotovideo has to know its version - so far we only track that in rpm :)

**Comment by aaannz** [comment]
Or just invent some not related to rpm version but API version?

**Comment by coolo** [comment]
hmm, we have an API towards the worker and we have an API towards the tests. What about a plain counter incremented with every breaking change and offer this through /isotovideo and through testapi?

**Comment by aaannz** [comment]
Something like that.

For different PR is however how worker should react. /isotovideo route is only available only when job is running, at that time worker is already known to openQA. Fail and unregister itself? Or send some notfication to openQA so admin can see outdated workers in worker page?

**Comment by coolo** [comment]
duplicate_job and stop itself.

**Comment by coolo** [comment]
now that I think about it, it only solves future problems as the current master worker won't check that route. So I assume this is actually a feature for after the merge ;)

**Comment by okurz** [comment]
better now than never?

yes, a plain counter. Maybe look [here](http://zguide.zeromq.org/page:all#UDP-Beacon-Framing) for an example how others are recommending the same. Might not be the best ref but the complete is actually a recommended read.

**Comment by coolo** [comment]
I have my own code to steal from ;)
https://github.com/icecc/icecream/blob/master/services/comm.h#L39

**Comment by coolo** [comment]
ok, I will merge these now - and then work on the version route in os-autoinst and we can add a check to the worker afterwards.

**Pull request merged**